.emacs.d/config.org
#+TITLE: Emacs configuration
#+DESCRIPTION: An org-babel based emacs configuration
#+LANGUAGE: en
#+PROPERTY: results silent
* Learning resources
|------------------------+-----------------------------------------------------------------------------|
| Link                   | description                                                                 |
|------------------------+-----------------------------------------------------------------------------|
| C-h t                  | Emacs tutorial                                                              |
| [[https://www.youtube.com/channel/UCDEtZ7AKmwS0_GNJog01D2g][Uncle Dave youtube]]     | Uncle Dave's youtube videos explain emacs configuration                     |
| [[https://www.youtube.com/channel/UCxkMDXQ5qzYOgXPRnOBrp1w][Mike Zamansky youtube]]  | Has a great set of videos on Emacs, computer science etc                    |
| [[https://www.udemy.com/course/getting-yourself-organized-with-org-mode/][Rainer Konig udemy]]     | Rainer Konig Udemy org-mode course, structured pdf and videos               |
| [[https://www.youtube.com/channel/UCfbGTpcJyEOMwKP-eYz3_fg][Rainer Konig youtube]]   | Youtube channel also from Rainer                                            |
| [[https://www.youtube.com/channel/UCQp2VLAOlvq142YN3JO3y8w][John Kitchin youtube]]   | This guy made org-ref and is an academic sensei                             |
| [[http://kitchingroup.cheme.cmu.edu/blog/category/emacs/1/][Kitchin research group]] | John Kitchin's research group site has a ton of resources emacs related     |
| C-h                    | This brings up a list with all types of help. E.g.C-h b describes bindings. |
| C-h c COMMAND          | Display name of the function (brief description of the command)             |
| C-h k COMMAND          | Display documentation of the function by typing the command                 |
| C-h f FUNCTION NAME    | Display documentation of the function by typing the function name           |
| C-h v                  | Display documentation of variables                                          |
| C-h a                  | Type keyword to list all commands whose names contain that keyword          |
| C-h i                  | Info - Read included manuals of installed packages                          |
|------------------------+-----------------------------------------------------------------------------|
* Explanation
So! You've found yourself on my personal emacs configuration files. First of all, welcome! To start, there are a total of three files needed for my configuration.

1. *init.el* is where the initial configuration of emacs begins. We initialize some stuff so we can use packages, require org, and load the other two important files!
2. *config.org* that is this file! Here we do some basic configuration that does not need a lot from your part. It is just my bare-bones emacs without fancy (and very usefull) extra stuff. The only thing to get started here, is to write down the path to the folder in which these three files live.
3. *config_custom.org* is where the magic happens. Well, extensive functionality that turns emacs from amazing to insane in the membrane! Here we also put shortcuts to other config files for quick access, we use org-journal, org-ref and org-roam (amazing combination). Basically everything that requires a bit more configuration on your part.

Have fun, and don't forget to adjust the emacs path!   
* File layout and paths
Some standard files for configuring. The only important thing here to edit is *emacs-dir* to the directory of your emacs configuration. Don't forget!!!
#+BEGIN_SRC emacs-lisp
  ;; General directories
  (setq emacs-dir "~/.emacs.d/")

  ;; Configuration files
  (setq emacs-config-org-file (concat emacs-dir "config.org"))
  (setq emacs-config-custom-org-file (concat emacs-dir "config_custom.org"))

  ;; More specific files
  (setq amx-items (concat emacs-dir "amx-items"))
  (setq org-reveal-root "http://cdn.jsdelivr.net/reveal.js/3.0.0/")

  ;; DOEN HET NOG NIET
  (setq backup-per-save (concat emacs-dir "backup/per-save"))
  (setq backup-per-session (concat emacs-dir "backup/per-session"))
  (setq libre-office-path "/Applications/LibreOffice.app/Contents/MacOS/soffice")
#+END_SRC
* Configuration
** Autocompletions
*** Company (autocompletion framework)
Company is a text completion framework for Emacs. The name stands for "complete anything". It uses pluggable back-ends and front-ends to retrieve and display completion candidates. See documentation on [[http://company-mode.github.io/][this site]].
#+BEGIN_SRC emacs-lisp
  (use-package company
    :ensure t
    :init
    (add-hook 'after-init-hook 'global-company-mode))
#+END_SRC
*** Electric pairing
Automatically pair the following elements on autocorrect. This is enabled in the [[*Standards (minor modes and minor improvements)][Standards (minor modes and minor improvements)]] through electric-pair-mode.

#+BEGIN_SRC emacs-lisp
  (setq electric-pair-pairs '(
                              (?\( . ?\))
                              (?\[ . ?\])
                              (?\" . ?\")
                              (?\{ . ?\})
                              (?\< . ?\>)
                              ))
#+END_SRC
*** IDO autocomplete filename searches
Ivy takes care of autocompleting filenames in the minibuffer at the bottom of the screen. It also keeps typing paths to a minimum as a folder is just an enter away.
#+BEGIN_SRC emacs-lisp
  (ivy-mode 1)
  (setq ivy-use-virtual-buffers t)
  (setq enable-recursive-minibuffers t)
  (global-set-key "\C-s" 'swiper)
#+END_SRC

Makes sure that M-x also generates suggestions. Otherwise you'd have to remember everything and not get autocompleted in M-x functions. Amx is the newer version of smex.
#+BEGIN_SRC emacs-lisp
  (use-package amx
    :ensure t
    :after ivy
    :custom
    (amx-backend 'auto)
    (ams-save-file amx-items)
    (amx-history-length 50)
    (amx-show-key-bindings nil)
    :config
    (amx-mode 1))
#+END_SRC
*** Yassnippet
Yassnippet is the templating system that is used. It creates a folder called snippets in which you can make a folder for each major mode you'd want a template for. E.g., python can have a few snippets to prettyfie a matplotlib graph, or org can have a template for exporting to a latex article or an html webpage.
#+BEGIN_SRC emacs-lisp
  (use-package yasnippet
    :ensure t
    :config (use-package yasnippet-snippets
              :ensure t)
    (yas-reload-all))
  (yas-global-mode 1)
#+END_SRC
** Buffers
All things buffer related
*** Buffer-move
Be able to swap buffers. See [[*Custom keystrokes][Custom keystrokes]] for shortcuts (buf-move-xxx).
#+BEGIN_SRC emacs-lisp
(use-package buffer-move
  :ensure t)
#+END_SRC
*** Ibuffer
Just a new buffer that lists the open buffers. It provides easy ways to close multiple buffers at once and navigate through them.
#+BEGIN_SRC emacs-lisp
  (global-set-key (kbd "C-x C-b") 'ibuffer)
  (setq ibuffer-expert t)
#+END_SRC
*** Killing buffers
**** Always kill current buffer
#+BEGIN_SRC emacs-lisp
  (global-set-key (kbd "C-x k") 'kill-current-buffer)
#+END_SRC
**** Kill all buffers
#+BEGIN_SRC emacs-lisp
  (defun kill-all-buffers ()
    (interactive)
    (mapc 'kill-buffer (buffer-list)))
  (global-set-key (kbd "C-M-s-k") 'kill-all-buffers)
#+END_SRC
*** Narrowing
Function to easily narrow and widen an area of code. If you select a piece of text, call this function, it will create a buffer with just that in it. This makes searching, or exporting just a part of something much easier.
#+BEGIN_SRC emacs-lisp
  (defun narrow-or-widen-dwim (p)
    "Widen if buffer is narrowed, narrow-dwim otherwise.
  Dwim means: region, org-src-block, org-subtree, or
  defun, whichever applies first. Narrowing to
  org-src-block actually calls `org-edit-src-code'.

  With prefix P, don't widen, just narrow even if buffer
  is already narrowed."
    (interactive "P")
    (declare (interactive-only))
    (cond ((and (buffer-narrowed-p) (not p)) (widen))
          ((region-active-p)
           (narrow-to-region (region-beginning)
                             (region-end)))
          ((derived-mode-p 'org-mode)
           ;; `org-edit-src-code' is not a real narrowing
           ;; command. Remove this first conditional if
           ;; you don't want it.
           (cond ((ignore-errors (org-edit-src-code) t)
                  (delete-other-windows))
                 ((ignore-errors (org-narrow-to-block) t))
                 (t (org-narrow-to-subtree))))
          ((derived-mode-p 'latex-mode)
           (LaTeX-narrow-to-environment))
          (t (narrow-to-defun))))
#+END_SRC
*** Switch to previous buffer
Small function to switch to previously used buffer.
#+BEGIN_SRC emacs-lisp
  (defun er-switch-to-previous-buffer ()
    "Switch to previously open buffer.
     Repeated invocations toggle between the two most recently open buffers."
    (interactive)
    (switch-to-buffer (other-buffer (current-buffer))))

  (global-set-key (kbd "C-c b") #'er-switch-to-previous-buffer)
#+END_SRC
*** switchwindow
Make switching buffer with C-x o easier. It provides you with shortcuts on the homerow to which buffer you want to go. Otherwise, you'd need to cycle through them which is aweful if you have multiple buffers on the screen.
#+BEGIN_SRC emacs-lisp
  (use-package switch-window
    :ensure t
    :config
    (setq switch-window-input-style 'minibuffer)
    (setq switch-window-increase 4)
    (setq switch-window-threshold 2)
    (setq switch-window-shortcut-style 'qwerty)
    (setq switch-window-qwerty-shortcuts
	  '("a" "s" "d" "f" "h" "j" "k" "l"))
    :bind
    ([remap other-window] . switch-window))
#+END_SRC
*** Toggle fullscreen buffer
When using multiple buffers at the same time, sometimes it's nice to toggle a single buffer as fullscreen.
#+BEGIN_SRC emacs-lisp
  (defun toggle-maximize-buffer () "Maximize buffer"
    (interactive)
    (if (= 1 (length (window-list)))
        (jump-to-register '_) 
      (progn
        (window-configuration-to-register '_)
        (delete-other-windows))))
  (global-set-key (kbd "C-M-f") 'toggle-maximize-buffer)
#+END_SRC
*** window splitting function
If you split the window into two buffers, follow the new buffer. You make a new one to work in there right?!
#+BEGIN_SRC emacs-lisp
  (defun split-and-follow-horizontally ()
    (interactive)
    (split-window-below)
    (balance-windows)
    (other-window 1))
  (global-set-key (kbd "C-x 2") 'split-and-follow-horizontally)

  (defun split-and-follow-vertically ()
    (interactive)
    (split-window-right)
    (balance-windows)
    (other-window 1))
  (global-set-key (kbd "C-x 3") 'split-and-follow-vertically)
#+END_SRC
** Custom functions
*** Backup files
Here we set where each file is backed up, how many versions of each file is backed
#+BEGIN_SRC emacs-lisp
  (setq version-control t     ;; Use version numbers for backups.
        kept-new-versions 10  ;; Number of newest versions to keep.
        kept-old-versions 0   ;; Number of oldest versions to keep.
        delete-old-versions t ;; Don't ask to delete excess backup versions.
        backup-by-copying t  ;; Copy all files, don't rename them.
        auto-save-interval 100 ;; Change interval of characters to which auto-save is enabled
        )

  (setq vc-make-backup-files t)

  ;; Default and per-save backups go here:
  (setq backup-directory-alist '(("" . "~/.emacs.d/backup/per-save")))

  (defun force-backup-of-buffer ()
    ;; Make a special "per session" backup at the first save of each
    ;; emacs session.
    (when (not buffer-backed-up)
      ;; Override the default parameters for per-session backups.
      (let ((backup-directory-alist '(("" . "~/.emacs.d/backup/per-session")))
            (kept-new-versions 3))
        (backup-buffer)))
    ;; Make a "per save" backup on each save.  The first save results in
    ;; both a per-session and a per-save backup, to keep the numbering
    ;; of per-save backups consistent.
    (let ((buffer-backed-up nil))
      (backup-buffer)))

  (add-hook 'before-save-hook  'force-backup-of-buffer)
#+END_SRC
*** Edit and reload config
Small function to easily configure and reload the configuration file.

#+BEGIN_SRC emacs-lisp
  (defun config-visit ()
    (interactive)
    (find-file emacs-config-org-file))

  (defun config-custom-visit ()
    (interactive)
    (find-file emacs-config-custom-org-file))

  (defun config-reload ()
    (interactive)
    (org-babel-load-file (expand-file-name emacs-config-org-file)))
#+END_SRC
** Custom keystrokes
All (most) the custom key combinations that I use regularly. It maps the Control z to its own map where I have most freedom.
#+BEGIN_SRC emacs-lisp
  ;; set up my own map for files, folder and windows
  (define-prefix-command 'z-map)
  (global-set-key (kbd "C-z") 'z-map)
  (define-key z-map (kbd "a") 'org-agenda-show-agenda-and-todo)
  (define-key z-map (kbd "c") 'avy-goto-char)
  (define-key z-map (kbd "n") 'narrow-or-widen-dwim)
  (define-key z-map (kbd "t") 'toggle-transparency)
  (define-key z-map (kbd "e") 'config-visit)
  (define-key z-map (kbd "E") 'config-custom-visit)
  (define-key z-map (kbd "r") 'config-reload)
  (define-key z-map (kbd "<left>") 'shrink-window-horizontally)
  (define-key z-map (kbd "<right>") 'enlarge-window-horizontally)
  (define-key z-map (kbd "<down>") 'shrink-window)
  (define-key z-map (kbd "<up>") 'enlarge-window)
  (define-key z-map (kbd "C-<up>") 'buf-move-up)
  (define-key z-map (kbd "C-<down>") 'buf-move-down)
  (define-key z-map (kbd "C-<left>") 'buf-move-left)
  (define-key z-map (kbd "C-<right>") 'buf-move-right)

  ;; ORG extra keybinding
  ;; Store a reference link to an org mode location
  (global-set-key (kbd "C-c l") 'org-store-link)

  ;; Add an extra cursor above or below current cursor
  (global-set-key (kbd "C-<") 'mc/mark-previous-like-this)
  (global-set-key (kbd "C->") 'mc/mark-next-like-this)

  ;; Remove an extra cursor above or below current cursor
  (global-set-key (kbd "C-,") 'mc/unmark-previous-like-this)
  (global-set-key (kbd "C-.") 'mc/unmark-next-like-this)

  ;; Skip a spot in adding a new cursor above or below
  (global-set-key (kbd "C-M-<") 'mc/skip-to-previous-like-this)
  (global-set-key (kbd "C-M->") 'mc/skip-to-next-like-this)

  ;; Mark all entries in current selection (useful if you want to rename a variable in the whole file)
  (global-set-key (kbd "C-M-,") 'mc/mark-all-like-this)

  ;; Create cursors on every line in selected area
  (global-set-key (kbd "C-M-.") 'mc/edit-lines)

  ;; Insert numbers with increased index for exery cursor (useful for lists)
  (global-set-key (kbd "C-;") 'mc/insert-numbers)

  ;; Same as numbers but then with letters
  (global-set-key (kbd "C-M-;") 'mc/insert-letters)

  ;; With control shift and a mouse-click add cursor
  (global-set-key (kbd "C-S-<mouse-1>") 'mc/add-cursor-on-click)
#+END_SRC
** Exporting
*** Org to latex blank lines
Here we make a small adaption in exporting to latex file. A double newline is translated to a bigskip, thus creating an extra whitespace in the resulting pdf.
#+BEGIN_SRC emacs-lisp
  ;; replace \n\n with bigskip
  (defun my-replace-double-newline (backend)
    "replace multiple blank lines with bigskip"
    (interactive)
    (goto-char (point-min))
    (while (re-search-forward "\\(^\\s-*$\\)\n\n+" nil t)
      (replace-match "\n#+LATEX: \\par\\vspace{\\baselineskip}\\noindent\n" nil t)
      ;;(replace-match "\n#+LATEX: \\bigskip\\noindent\n" nil t)
      (forward-char 1)))

  (add-hook 'org-export-before-processing-hook 'my-replace-double-newline)
#+END_SRC
*** Export to word
Make sure that export (C-e) to odt, will be formatted to a .doc document for word.
#+BEGIN_SRC emacs-lisp
  ;; This setup is tested on Emacs 24.3 & Emacs 24.4 on Linux/OSX
  ;; org v7 bundled with Emacs 24.3
  (setq org-export-odt-preferred-output-format "doc")
  ;; org v8 bundled with Emacs 24.4
  (setq org-odt-preferred-output-format "doc")
  ;; BTW, you can assign "pdf" in above variables if you prefer PDF format

  ;; Only OSX need below setup
  (defun my-setup-odt-org-convert-process ()
    (setq process-string "/Applications/LibreOffice.app/Contents/MacOS/soffice --headless --convert-to %f%x --outdir %d %i")
    (interactive)
    (let ((cmd libre-office-path))
      (when (and (eq system-type 'darwin) (file-exists-p cmd))
	;; org v7
	(setq org-export-odt-convert-processes '(("LibreOffice" "/Applications/LibreOffice.app/Contents/MacOS/soffice --headless --convert-to %f%x --outdir %d %i")))
	;; org v8
	(setq org-odt-convert-processes '(("LibreOffice"  "/Applications/LibreOffice.app/Contents/MacOS/soffice --headless --convert-to %f%x --outdir %d %i"))))
      ))
  (my-setup-odt-org-convert-process)
#+END_SRC
*** Reveal.js
Provide the option to export (C-e) an org-file to a reveal presentation.
#+BEGIN_SRC emacs-lisp
  (use-package ox-reveal
  :ensure ox-reveal)
  (setq org-reveal-mathjax t)
  (use-package htmlize :ensure t)
#+END_SRC
*** Export to subdirectory
Exporting brings about a lot of extra files and mess in the folder of your org file. This variable basically deletes all the extra trash that is made in order to provide you with a pdf file. Keep things nice and clean! :D
#+BEGIN_SRC emacs-lisp
  (setq org-latex-logfiles-extensions (quote ("lof" "lot" "tex~" "aux" "idx" "log" "out" "toc" "nav" "snm" "vrb" "dvi" "fdb_latexmk" "blg" "brf" "fls" "entoc" "ps" "spl" "bbl")))
#+END_SRC
*** Beamer
#+BEGIN_SRC emacs-lisp
  (require 'ox-beamer)
#+END_SRC
** Gimmicks
Just some small functions that can be used for (almost) useless things.
*** Transparency
#+BEGIN_SRC emacs-lisp
  ;;(set-frame-parameter (selected-frame) 'alpha '(<active> . <inactive>))
  ;;(set-frame-parameter (selected-frame) 'alpha <both>)
  (set-frame-parameter (selected-frame) 'alpha '(100 . 100))
  (add-to-list 'default-frame-alist '(alpha . (100 . 100)))

  (defun toggle-transparency ()
     (interactive)
     (let ((alpha (frame-parameter nil 'alpha)))
       (set-frame-parameter

.emacs.d/config.el
;; General directories
(setq emacs-dir "~/.emacs.d/")

;; Configuration files
(setq emacs-config-org-file (concat emacs-dir "config.org"))
(setq emacs-config-custom-org-file (concat emacs-dir "config_custom.org"))

;; More specific files
(setq amx-items (concat emacs-dir "amx-items"))
(setq org-reveal-root "http://cdn.jsdelivr.net/reveal.js/3.0.0/")

;; DOEN HET NOG NIET
(setq backup-per-save (concat emacs-dir "backup/per-save"))
(setq backup-per-session (concat emacs-dir "backup/per-session"))
(setq libre-office-path "/Applications/LibreOffice.app/Contents/MacOS/soffice")

(use-package company
  :ensure t
  :init
  (add-hook 'after-init-hook 'global-company-mode))

(setq electric-pair-pairs '(
                            (?\( . ?\))
                            (?\[ . ?\])
                            (?\" . ?\")
                            (?\{ . ?\})
                            (?\< . ?\>)
                            ))

(ivy-mode 1)
(setq ivy-use-virtual-buffers t)
(setq enable-recursive-minibuffers t)
(global-set-key "\C-s" 'swiper)

(use-package amx
  :ensure t
  :after ivy
  :custom
  (amx-backend 'auto)
  (ams-save-file amx-items)
  (amx-history-length 50)
  (amx-show-key-bindings nil)
  :config
  (amx-mode 1))

(use-package yasnippet
  :ensure t
  :config (use-package yasnippet-snippets
            :ensure t)
  (yas-reload-all))
(yas-global-mode 1)

(use-package buffer-move
  :ensure t)

(global-set-key (kbd "C-x C-b") 'ibuffer)
(setq ibuffer-expert t)

(global-set-key (kbd "C-x k") 'kill-current-buffer)

(defun kill-all-buffers ()
  (interactive)
  (mapc 'kill-buffer (buffer-list)))
(global-set-key (kbd "C-M-s-k") 'kill-all-buffers)

(defun narrow-or-widen-dwim (p)
  "Widen if buffer is narrowed, narrow-dwim otherwise.
Dwim means: region, org-src-block, org-subtree, or
defun, whichever applies first. Narrowing to
org-src-block actually calls `org-edit-src-code'.

With prefix P, don't widen, just narrow even if buffer
is already narrowed."
  (interactive "P")
  (declare (interactive-only))
  (cond ((and (buffer-narrowed-p) (not p)) (widen))
        ((region-active-p)
         (narrow-to-region (region-beginning)
                           (region-end)))
        ((derived-mode-p 'org-mode)
         ;; `org-edit-src-code' is not a real narrowing
         ;; command. Remove this first conditional if
         ;; you don't want it.
         (cond ((ignore-errors (org-edit-src-code) t)
                (delete-other-windows))
               ((ignore-errors (org-narrow-to-block) t))
               (t (org-narrow-to-subtree))))
        ((derived-mode-p 'latex-mode)
         (LaTeX-narrow-to-environment))
        (t (narrow-to-defun))))

(defun er-switch-to-previous-buffer ()
  "Switch to previously open buffer.
   Repeated invocations toggle between the two most recently open buffers."
  (interactive)
  (switch-to-buffer (other-buffer (current-buffer))))

(global-set-key (kbd "C-c b") #'er-switch-to-previous-buffer)

(use-package switch-window
  :ensure t
  :config
  (setq switch-window-input-style 'minibuffer)
  (setq switch-window-increase 4)
  (setq switch-window-threshold 2)
  (setq switch-window-shortcut-style 'qwerty)
  (setq switch-window-qwerty-shortcuts
	'("a" "s" "d" "f" "h" "j" "k" "l"))
  :bind
  ([remap other-window] . switch-window))

(defun toggle-maximize-buffer () "Maximize buffer"
  (interactive)
  (if (= 1 (length (window-list)))
      (jump-to-register '_) 
    (progn
      (window-configuration-to-register '_)
      (delete-other-windows))))
(global-set-key (kbd "C-M-f") 'toggle-maximize-buffer)

(defun split-and-follow-horizontally ()
  (interactive)
  (split-window-below)
  (balance-windows)
  (other-window 1))
(global-set-key (kbd "C-x 2") 'split-and-follow-horizontally)

(defun split-and-follow-vertically ()
  (interactive)
  (split-window-right)
  (balance-windows)
  (other-window 1))
(global-set-key (kbd "C-x 3") 'split-and-follow-vertically)

(setq version-control t     ;; Use version numbers for backups.
      kept-new-versions 10  ;; Number of newest versions to keep.
      kept-old-versions 0   ;; Number of oldest versions to keep.
      delete-old-versions t ;; Don't ask to delete excess backup versions.
      backup-by-copying t  ;; Copy all files, don't rename them.
      auto-save-interval 100 ;; Change interval of characters to which auto-save is enabled
      )

(setq vc-make-backup-files t)

;; Default and per-save backups go here:
(setq backup-directory-alist '(("" . "~/.emacs.d/backup/per-save")))

(defun force-backup-of-buffer ()
  ;; Make a special "per session" backup at the first save of each
  ;; emacs session.
  (when (not buffer-backed-up)
    ;; Override the default parameters for per-session backups.
    (let ((backup-directory-alist '(("" . "~/.emacs.d/backup/per-session")))
          (kept-new-versions 3))
      (backup-buffer)))
  ;; Make a "per save" backup on each save.  The first save results in
  ;; both a per-session and a per-save backup, to keep the numbering
  ;; of per-save backups consistent.
  (let ((buffer-backed-up nil))
    (backup-buffer)))

(add-hook 'before-save-hook  'force-backup-of-buffer)

(defun config-visit ()
  (interactive)
  (find-file emacs-config-org-file))

(defun config-custom-visit ()
  (interactive)
  (find-file emacs-config-custom-org-file))

(defun config-reload ()
  (interactive)
  (org-babel-load-file (expand-file-name emacs-config-org-file)))

;; set up my own map for files, folder and windows
(define-prefix-command 'z-map)
(global-set-key (kbd "C-z") 'z-map)
(define-key z-map (kbd "a") 'org-agenda-show-agenda-and-todo)
(define-key z-map (kbd "c") 'avy-goto-char)
(define-key z-map (kbd "n") 'narrow-or-widen-dwim)
(define-key z-map (kbd "t") 'toggle-transparency)
(define-key z-map (kbd "e") 'config-visit)
(define-key z-map (kbd "E") 'config-custom-visit)
(define-key z-map (kbd "r") 'config-reload)
(define-key z-map (kbd "<left>") 'shrink-window-horizontally)
(define-key z-map (kbd "<right>") 'enlarge-window-horizontally)
(define-key z-map (kbd "<down>") 'shrink-window)
(define-key z-map (kbd "<up>") 'enlarge-window)
(define-key z-map (kbd "C-<up>") 'buf-move-up)
(define-key z-map (kbd "C-<down>") 'buf-move-down)
(define-key z-map (kbd "C-<left>") 'buf-move-left)
(define-key z-map (kbd "C-<right>") 'buf-move-right)

;; ORG extra keybinding
;; Store a reference link to an org mode location
(global-set-key (kbd "C-c l") 'org-store-link)

;; Add an extra cursor above or below current cursor
(global-set-key (kbd "C-<") 'mc/mark-previous-like-this)
(global-set-key (kbd "C->") 'mc/mark-next-like-this)

;; Remove an extra cursor above or below current cursor
(global-set-key (kbd "C-,") 'mc/unmark-previous-like-this)
(global-set-key (kbd "C-.") 'mc/unmark-next-like-this)

;; Skip a spot in adding a new cursor above or below
(global-set-key (kbd "C-M-<") 'mc/skip-to-previous-like-this)
(global-set-key (kbd "C-M->") 'mc/skip-to-next-like-this)

;; Mark all entries in current selection (useful if you want to rename a variable in the whole file)
(global-set-key (kbd "C-M-,") 'mc/mark-all-like-this)

;; Create cursors on every line in selected area
(global-set-key (kbd "C-M-.") 'mc/edit-lines)

;; Insert numbers with increased index for exery cursor (useful for lists)
(global-set-key (kbd "C-;") 'mc/insert-numbers)

;; Same as numbers but then with letters
(global-set-key (kbd "C-M-;") 'mc/insert-letters)

;; With control shift and a mouse-click add cursor
(global-set-key (kbd "C-S-<mouse-1>") 'mc/add-cursor-on-click)

;; replace \n\n with bigskip
(defun my-replace-double-newline (backend)
  "replace multiple blank lines with bigskip"
  (interactive)
  (goto-char (point-min))
  (while (re-search-forward "\\(^\\s-*$\\)\n\n+" nil t)
    (replace-match "\n#+LATEX: \\par\\vspace{\\baselineskip}\\noindent\n" nil t)
    ;;(replace-match "\n#+LATEX: \\bigskip\\noindent\n" nil t)
    (forward-char 1)))

(add-hook 'org-export-before-processing-hook 'my-replace-double-newline)

;; This setup is tested on Emacs 24.3 & Emacs 24.4 on Linux/OSX
;; org v7 bundled with Emacs 24.3
(setq org-export-odt-preferred-output-format "doc")
;; org v8 bundled with Emacs 24.4
(setq org-odt-preferred-output-format "doc")
;; BTW, you can assign "pdf" in above variables if you prefer PDF format

;; Only OSX need below setup
(defun my-setup-odt-org-convert-process ()
  (setq process-string "/Applications/LibreOffice.app/Contents/MacOS/soffice --headless --convert-to %f%x --outdir %d %i")
  (interactive)
  (let ((cmd libre-office-path))
    (when (and (eq system-type 'darwin) (file-exists-p cmd))
      ;; org v7
      (setq org-export-odt-convert-processes '(("LibreOffice" "/Applications/LibreOffice.app/Contents/MacOS/soffice --headless --convert-to %f%x --outdir %d %i")))
      ;; org v8
      (setq org-odt-convert-processes '(("LibreOffice"  "/Applications/LibreOffice.app/Contents/MacOS/soffice --headless --convert-to %f%x --outdir %d %i"))))
    ))
(my-setup-odt-org-convert-process)

(use-package ox-reveal
:ensure ox-reveal)
(setq org-reveal-mathjax t)
(use-package htmlize :ensure t)

(setq org-latex-logfiles-extensions (quote ("lof" "lot" "tex~" "aux" "idx" "log" "out" "toc" "nav" "snm" "vrb" "dvi" "fdb_latexmk" "blg" "brf" "fls" "entoc" "ps" "spl" "bbl")))

(require 'ox-beamer)

;;(set-frame-parameter (selected-frame) 'alpha '(<active> . <inactive>))
;;(set-frame-parameter (selected-frame) 'alpha <both>)
(set-frame-parameter (selected-frame) 'alpha '(100 . 100))
(add-to-list 'default-frame-alist '(alpha . (100 . 100)))

(defun toggle-transparency ()
   (interactive)
   (let ((alpha (frame-parameter nil 'alpha)))
     (set-frame-parameter
      nil 'alpha
      (if (eql (cond ((numberp alpha) alpha)
                     ((numberp (cdr alpha)) (cdr alpha))
                     ;; Also handle undocumented (<active> <inactive>) form.
                     ((numberp (cadr alpha)) (cadr alpha)))
               100)
          '(95 . 95) '(100 . 100)))))

(use-package avy :ensure t)

(require 'multiple-cursors)

(use-package org-bullets
  :ensure t
  :config
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode))))

(setq org-indent-indentation-per-level 1)             ;; Shorten the space on the left side with org headers
(setq org-adapt-indentation nil)                      ;; Adapt indentation to outline node level. Set to nill as it takes up space.
(setq org-hide-emphasis-markers t)                    ;; When making something bold *Hallo*, hide stars. Goes for all emphasis markers.
(setq org-cycle-separator-lines 1)                    ;; Leave a single empty line between headers if there is one. Otherwise leave no room or make the empty lines belong to the previous header.
(setq org-hide-leading-stars 't)                      ;; Hide the extra stars in front of a header (org-bullet displays nicer, but why add extra package)
(customize-set-variable 'org-blank-before-new-entry 
                        '((heading . nil)
                          (plain-list-item . nil)))   ;; Dont randomly remove newlines below headers

(setq org-image-actual-width 600)

(org-babel-do-load-languages
 'org-babel-load-languages
 '((python . t)))

(setq org-emphasis-alist
      '(("*" bold)
	("/" italic)
	("_" underline)
	("=" (:foreground "#ff8fdd"))
	("~" (:foreground "#ffb957"))
	("+" (:strike-through t))))

(font-lock-add-keywords
 'org-mode
 '(("\\(!_[^!\n]+_!\\)" (0 '(:foreground "#9afcae") t))))

(font-lock-add-keywords
 'org-mode
 '(("\\(&[^&\n]+&\\)" (0 '(:foreground "#85b4ff") t))))

(setq save-interprogram-paste-before-kill t) ;; Perpetuates system clipboard
(setq scroll-conservatively 1)      ;; Keep from making huge jumps when scrolling
(setq ring-bell-function 'ignore)   ;; Unable annoying sounds
(setq visible-bell 1)               ;; disable annoying windows sound
(setq inhibit-startup-message t)    ;; Hide the startup message
(setq display-time-24hr-format t)   ;; Format clock
(setq-default display-line-numbers 'relative) ;; Setting the line numbers
(when window-system (global-hl-line-mode t)) ;; Get a current line shadow in IDE
(defalias 'yes-or-no-p 'y-or-n-p)   ;; Replace yes questions to y
(setq sentence-end-double-space nil);; Start a new sentence with just a single space instead of 2
(exec-path-from-shell-initialize)   ;; Fixes path issues on mac emacs.app so that same env of terminal is used

(use-package hungry-delete
  :ensure t
  :config (global-hungry-delete-mode))

(tool-bar-mode -1)                  ;; Get rid of tool-bar
(menu-bar-mode -1)                  ;; Git rid of menu
(scroll-bar-mode -1)                ;; Get rid of scroll-bar
(global-auto-revert-mode 1)         ;; Make sure that you're always looking at the latest version of a file. Change file when changed on disk
(delete-selection-mode 1)           ;; Remove text from selection instead of just inserting text
(display-time-mode 1)               ;; Set clock on lower right side
(electric-pair-mode t)              ;; Enable electric pair mode. It autocompletes certain pairs. E.g., (), {}, [], <>
(global-subword-mode 1)             ;; Cause M-f to move forward per capitalization within a word. E.g., weStopAtEveryCapital
(global-visual-line-mode 1)                ;; Make sure that lines do not disapear at the right side of the screen but wrap around

(use-package popup-kill-ring
  :ensure t
  :bind ("M-y" . popup-kill-ring))

(use-package which-key
  :ensure t
  :init
  (which-key-mode))

(setq org-todo-keyword-faces
      '(
        ("DOING" . (:foreground "#05d3fc" :weight bold :box (:line-width 2 :style released-button)))
        ("WAITING" . (:foreground "#fcca05" :weight bold :box (:line-width 2 :style released-button)))
        ("FLEETING" . (:foreground "#f62af9" :weight bold :box (:line-width 2 :style released-button)))
        ("LONGTERM" . (:foreground "#c4013c" :weight bold :box (:line-width 2 :style released-button)))
        ("CANCELED" . (:foreground "#fc4205" :weight bold :box (:line-width 2 :style released-button)))
        ))

(setq org-todo-keywords
      '((sequence "TODO(t)" "DOING(d)" "WAITING(w)" "FLEETING(f)" "|" "LONGTERM(l)" "CANCELED(c)" "DONE(f)")))

(use-package try
    :ensure t)

(use-package beacon
    :ensure t
    :init
    (beacon-mode 1))

(use-package rainbow-mode
  :ensure t
  :init (add-hook 'prog-mode-hook 'rainbow-mode))

(use-package rainbow-delimiters
  :ensure t
  :init
  (rainbow-delimiters-mode 1))

(use-package doom-themes
  :ensure t
  :config
  ;; Global settings (defaults)
  (setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
	doom-themes-enable-italic t) ; if nil, italics is universally disabled
  (load-theme 'doom-one t)

  ;; Enable flashing mode-line on errors
  (doom-themes-visual-bell-config)

  ;; Enable custom neotree theme (all-the-icons must be installed!)
  (doom-themes-neotree-config)
  ;; or for treemacs users
  (setq doom-themes-treemacs-theme "doom-colors") ; use the colorful treemacs theme
  (doom-themes-treemacs-config)

  ;; Corrects (and improves) org-mode's native fontification.
  (doom-themes-org-config))

(use-package doom-modeline
  :ensure t
  :init (doom-modeline-mode 1))

;; show wordcount inside modeline
(setq doom-modeline-enable-word-count t)

(use-package all-the-icons :ensure t)
